### ISSUE TITLE
limit(floor(n*x),x,0) for n > 10^8

### ISSUE THREAD

**Issue by rtoy** [rtoy/maxima#2255]
Imported from SourceForge on 2024-07-05 22:35:06
Created by **[willisbl](https://sourceforge.net/u/willisbl/)** on 2016-09-01 10:31:53
Original: https://sourceforge.net/p/maxima/bugs/3203

---

OK:
    (%i1) limit(floor(10^8*x),x,0,minus);
    (%o1) -1

Suboptimal:
    (%i2) limit(floor((10^8 +1)*x),x,0,minus);
    (%o2) limit(floor(100000001*x),x,0,minus)
    
The reason for the switch at 10^8 is that the limit code uses variable assumptions that involve 10^8.

A fix is to give floor and ceiling a simplim%function

~~~
(setf (get '$floor 'simplim%function)  'simplim-floor)

(defun simplim-floor (e x pt) ; pt is either zeroa or zerob
  (let  ((preserve-direction nil))
    (setq e ($limit (cadr e) x 0 (if (eq pt '$zerob) '$minus '$plus))) ;limit of argument to floor
    (cond ((not ($featurep e '$real)) (throw 'limit nil))
	  (($featurep e '$noninteger)  (take '($floor) e)) ;floor is continuous away from an integer
	  (($featurep e '$integer)  
	   (add e (if (eq pt '$zerob) -1 0)))
	  (t (throw 'limit nil))))) 
	   
(setf (get '$ceiling 'simplim%function)  'simplim-ceiling)

(defun simplim-ceiling (e x pt) ; pt is either zeroa or zerob
  (let  ((preserve-direction nil))
    (setq e ($limit (cadr e) x 0 (if (eq pt '$zerob) '$minus '$plus))) ;limit of argument to floor
    (cond ((not ($featurep e '$real)) (throw 'limit nil))
	  (($featurep e '$noninteger)  (take '($ceiling) e)) ;ceiling is continuous away from an integer
	  (($featurep e '$integer)  
	   (add e (if (eq pt '$zerob) 0 1)))
	  (t (throw 'limit nil))))) 
~~~

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-05 22:35:07
Created by **[dgildea](https://sourceforge.net/u/dgildea/)** on 2016-09-05 08:45:43
Original: https://sourceforge.net/p/maxima/bugs/3203/#d168

---

- **status**: open --> closed
- **assigned_to**: Dan Gildea

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-05 22:35:11
Created by **[dgildea](https://sourceforge.net/u/dgildea/)** on 2016-09-05 08:45:43
Original: https://sourceforge.net/p/maxima/bugs/3203/#69b7

---

fixed in commit [c54f5339c543f3e32ba8a106f5e6b1e20d756ba9]
